consul/server_manager/server_manager_test.go

```diff
@@ -0,0 +1,333 @@
+)
+
+var (
+	localLogger    *log.Logger
+	localLogBuffer *bytes.Buffer
+)
+
+func init() {
+	localLogBuffer = new(bytes.Buffer)
+	localLogger = log.New(localLogBuffer, "", 0)
+}
+
+func GetBufferedLogger() *log.Logger {
+	return localLogger
+}
+
+type fauxSerf struct {
+}
+
+func (s *fauxSerf) NumNodes() int {
+	return 16384
+}
+
+func testServerManager() (sm *server_manager.ServerManager) {
+	logger := GetBufferedLogger()
+	logger = log.New(os.Stderr, "", log.LstdFlags)
+	shutdownCh := make(chan struct{})
+	sm = server_manager.New(logger, shutdownCh, &fauxSerf{})
+	return sm
+}
+
+// func (sm *ServerManager) AddServer(server *server_details.ServerDetails) {
+func TestServerManager_AddServer(t *testing.T) {
+	sm := testServerManager()
+	var num int
+	num = sm.NumServers()
+	if num != 0 {
+		t.Fatalf("Expected zero servers to start")
+	}
+
+	s1 := &server_details.ServerDetails{Name: "s1"}
+	sm.AddServer(s1)
+	num = sm.NumServers()
+	if num != 1 {
+		t.Fatalf("Expected one server")
+	}
+
+	sm.AddServer(s1)
+	num = sm.NumServers()
+	if num != 1 {
+		t.Fatalf("Expected one server (still)")
+	}
+
+	s2 := &server_details.ServerDetails{Name: "s2"}
+	sm.AddServer(s2)
+	num = sm.NumServers()
+	if num != 2 {
+		t.Fatalf("Expected two servers")
+	}
+}
+
+// func (sm *ServerManager) FindServer() (server *server_details.ServerDetails) {
+func TestServerManager_FindServer(t *testing.T) {
+	sm := testServerManager()
+
+	if sm.FindServer() != nil {
+		t.Fatalf("Expected nil return")
+	}
+
+	sm.AddServer(&server_details.ServerDetails{Name: "s1"})
+	if sm.NumServers() != 1 {
+		t.Fatalf("Expected one server")
+	}
+
+	s1 := sm.FindServer()
+	if s1 == nil {
+		t.Fatalf("Expected non-nil server")
+	}
+	if s1.Name != "s1" {
+		t.Fatalf("Expected s1 server")
+	}
+
+	s1 = sm.FindServer()
+	if s1 == nil || s1.Name != "s1" {
+		t.Fatalf("Expected s1 server (still)")
+	}
+
+	sm.AddServer(&server_details.ServerDetails{Name: "s2"})
+	if sm.NumServers() != 2 {
+		t.Fatalf("Expected two servers")
+	}
+	s1 = sm.FindServer()
+	if s1 == nil || s1.Name != "s1" {
+		t.Fatalf("Expected s1 server (still)")
+	}
+
+	sm.NotifyFailedServer(s1)
+	s2 := sm.FindServer()
+	if s2 == nil || s2.Name != "s2" {
+		t.Fatalf("Expected s2 server")
+	}
+
+	sm.NotifyFailedServer(s2)
+	s1 = sm.FindServer()
+	if s1 == nil || s1.Name != "s1" {
+		t.Fatalf("Expected s1 server")
+	}
+}
+
+// func New(logger *log.Logger, shutdownCh chan struct{}) (sm *ServerManager) {
+func TestServerManager_New(t *testing.T) {
+	logger := GetBufferedLogger()
+	logger = log.New(os.Stderr, "", log.LstdFlags)
+	shutdownCh := make(chan struct{})
+	sm := server_manager.New(logger, shutdownCh, &fauxSerf{})
+	if sm == nil {
+		t.Fatalf("ServerManager nil")
+	}
+}
+
+// func (sm *ServerManager) NotifyFailedServer(server *server_details.ServerDetails) {
+func TestServerManager_NotifyFailedServer(t *testing.T) {
+	sm := testServerManager()
+
+	if sm.NumServers() != 0 {
+		t.Fatalf("Expected zero servers to start")
+	}
+
```

Comment:
call this with no servers in the list, as well as a bogus server that doesn't exist in the list once there are some in there